src/ranged.cpp

```diff
@@ -376,6 +374,69 @@ dealt_projectile_attack Creature::projectile_attack( const projectile &proj_arg,
     return attack;
 }
 
+double player::gun_engagement_range( const item& gun, int aim, int penalty, unsigned chance, double accuracy ) const
+{
+    if( !gun.is_gun() || !gun.ammo_sufficient() ) {
+        return 0;
+    }
+
+    if( chance == 0 ) {
+        return gun.gun_range( this );
+    }
+
+    int driving = 0;
+    if( penalty < 0 ) {
+        // get current effective player recoil
+        penalty = std::max( recoil, 0 );
+        if( recoil < 0 ) {
+            debugmsg( "negative player recoil when calculating effective range" );
+        }
+
+        // get maximum penalty from driving
+        driving = std::max( driving_recoil, 0 );
+        if( driving_recoil < 0 ) {
+            debugmsg( "negative driving recoil when calculating effective range" );
+        }
+    }
+
+    // aim_per_time() returns improvement (in MOA) per 10 moves
+    for( int i = aim * 10; i != 0; --i ) {
+        int adj = aim_per_time( gun, penalty );
+        if( adj <= 0 ) {
+            break; // no further improvement is possible
+        }
+        penalty -= adj;
+    }
+
+    // calculate maximum potential dispersion
+    double dispersion = get_weapon_dispersion( &gun, false ) + penalty + driving;
+
+    // dispersion is uniformly distributed at random so scale linearly with chance
```

Comment:
The comment doesn't match the current behavior but I'd like to change the latter as it's very hard to write deterministic unit tests for a non-uniform distribution.

For the current code the random value is ~57% of the maximum at `MIN_RECOIL` rising to ~72% at `MIN_RECOIL * 3` (both calculated via 10⁴ iterations). Any changes within `player::get_weapon_dispersion` will affect this distribution.

I'd like to separate out weapon and player recoil to be returned separately before randomness is applied so that the unit tests can be deterministic. A weighted rng can still be used providing it's bias is known.